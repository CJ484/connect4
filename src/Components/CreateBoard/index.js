import { findLastIndex } from "lodash";
import gameSetting from "../../const/gameSettings";

  const Board = ({winner, board, setBoard, currentPlayer}) => {
    const numRows = gameSetting.numRows;

    
  const handleCellClick = (col) => {
    if (winner) return;
    const updatedGrid = board.map((row) => [...row]);
    const emptyRow = findLastIndex(updatedGrid, (row) => !row[col]);

    if (emptyRow !== -1) {
      updatedGrid[emptyRow][col] = currentPlayer;
      setBoard(updatedGrid);
    }
  };

      const renderCell = (row, col) => {
        const cellLocation = board[row][col];
        return (
          <div
            key={col}
            className={`cell ${cellLocation}`}
            onClick={() => handleCellClick(col)}
          />
        );
      };

      const renderRow = (row) => {
        return (
          <div key={row} id={`row${row}`} className="row">
            {board[row].map((_, col) => (
              <div key={col} className="col">
                {renderCell(row, col)}
              </div>
            ))}
          </div>
        );
      };

    return (
      <div className="game-board">
        {Array.from({ length: numRows }, (_, row) => renderRow(row))}
      </div>
    );
}

export default Board;