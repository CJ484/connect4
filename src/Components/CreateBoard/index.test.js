import React from "react";
import { fireEvent, render, screen} from "@testing-library/react";
import '@testing-library/jest-dom';
import ReactTestRenderer from "react-test-renderer";
import CreateBoard from "./index";
import gameSetting from "../../const/gameSettings";

jest.mock("../../const/gameSettings", () => ({
  numRows: 3,
}));

describe("CreateBoard component", () => {
  let setBoardMock;
  let board;
  let props;

  beforeEach(() => {
    setBoardMock = jest.fn();
    board = [
      [null, null, null],
      [null, null, null],
      [null, null, null],
    ];
    props = {
      winner: null,
      board,
      setBoard: setBoardMock,
      currentPlayer: "X",
    };
  });

  test("renders the correct number of rows and cells", () => {
    render(<CreateBoard {...props} />);
    const rows = screen.getAllByTestId("row");
    expect(rows).toHaveLength(gameSetting.numRows); // Change this to match `gameSetting.numRows`

    const cells = screen.getAllByTestId("cell");
    expect(cells).toHaveLength(3 * 3);
  });


  it("updates the board when a cell is clicked", () => {
    render(<CreateBoard {...props} />);
    fireEvent.click(screen.getAllByTestId("cell")[0]);
    expect(setBoardMock).toHaveBeenCalledTimes(1);

    const expectedBoard = [
      [null, null, null],
      [null, null, null],
      ["X", null, null],
    ];
    expect(setBoardMock).toHaveBeenCalledWith(expectedBoard);
  });

  it("does not update the board when there is a winner", () => {
    props.winner = "X";
    render(<CreateBoard {...props}/>)
    fireEvent.click(screen.getAllByTestId("cell")[1]);
    expect(setBoardMock).toHaveBeenCalledTimes(0);
  });

  it("renders cells with correct class names based on cell content", () => {
    props.board = [
      [null, "O", null],
      ["X", null, null],
      ["O", "X", "O"],
    ];
    render(<CreateBoard {...props} />);
    const cells = screen.getAllByTestId("cell");
    expect(cells[0]).toHaveClass("null");
    expect(cells[1]).toHaveClass("O");
    expect(cells[3]).toHaveClass("X");
    expect(cells[6]).toHaveClass("O");
  });

  it('matches the snapshot', () => {
    const component = ReactTestRenderer.create(<CreateBoard {...props}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
