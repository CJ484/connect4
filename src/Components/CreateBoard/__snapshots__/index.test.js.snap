// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBoard component matches the snapshot 1`] = `
<div
  className="game-board"
>
  <div
    className="row"
    data-testid="row"
  >
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="row"
    data-testid="row"
  >
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="row"
    data-testid="row"
  >
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
    <div
      className="col"
    >
      <div
        className="cell null"
        data-testid="cell"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
